File: macos/DigipalPlayer/DigipalPlayer/Services/MediaManager.swift
import Foundation

class MediaManager {
    static let shared = MediaManager()

    private let mediaDir: URL
    private let manifestPath: URL
    private var manifest: [String: MediaEntry] = [:]
    private var activeDownloads: Set<String> = []
    private let maxConcurrent = 3
    private let lowStorageThreshold: UInt64 = 100 * 1024 * 1024
    private let queue = DispatchQueue(label: "com.digipal.player.media", qos: .utility)

    struct MediaEntry: Codable {
        var localPath: String
        var size: Int64
        var downloadedAt: Double
        var lastAccessed: Double
    }

    private init() {
        let appSupport = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
        let playerDir = appSupport.appendingPathComponent("DigipalPlayer")
        self.mediaDir = playerDir.appendingPathComponent("media")
        self.manifestPath = playerDir.appendingPathComponent("media-manifest.json")

        try? FileManager.default.createDirectory(at: mediaDir, withIntermediateDirectories: true)
        loadManifest()
        cleanupOrphans()
    }

    private func loadManifest() {
        guard FileManager.default.fileExists(atPath: manifestPath.path) else { return }
        do {
            let data = try Data(contentsOf: manifestPath)
            manifest = try JSONDecoder().decode([String: MediaEntry].self, from: data)
        } catch {
            print("[MediaManager] Failed to load manifest: \(error.localizedDescription)")
            manifest = [:]
        }
    }

    private func saveManifest() {
        do {
            let data = try JSONEncoder().encode(manifest)
            try data.write(to: manifestPath)
        } catch {
            print("[MediaManager] Failed to save manifest: \(error.localizedDescription)")
        }
    }

    private func objectPathToFilename(_ objectPath: String) -> String {
        return objectPath
            .replacingOccurrences(of: "/objects/", with: "")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "\\", with: "_")
            .replacingOccurrences(of: ":", with: "_")
            .replacingOccurrences(of: "*", with: "_")
            .replacingOccurrences(of: "?", with: "_")
            .replacingOccurrences(of: "\"", with: "_")
            .replacingOccurrences(of: "<", with: "_")
            .replacingOccurrences(of: ">", with: "_")
            .replacingOccurrences(of: "|", with: "_")
    }

    func getLocalMediaPath(objectPath: String) -> String {
        guard var entry = manifest[objectPath] else { return "" }
        let filePath = entry.localPath
        guard FileManager.default.fileExists(atPath: filePath) else {
            manifest.removeValue(forKey: objectPath)
            saveManifest()
            return ""
        }
        entry.lastAccessed = Date().timeIntervalSince1970 * 1000
        manifest[objectPath] = entry
        saveManifest()
        return "local-media://" + filePath.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed)!
    }

    func getAllLocalPaths() -> [(String, String)] {
        return manifest.compactMap { (objectPath, entry) in
            guard FileManager.default.fileExists(atPath: entry.localPath) else { return nil }
            let protocolUrl = "local-media://" + (entry.localPath.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? entry.localPath)
            return (objectPath, protocolUrl)
        }
    }

    func downloadMedia(objectPath: String, signedUrl: String, completion: @escaping (String) -> Void) {
        queue.async { [weak self] in
            guard let self = self else { return }

            if self.activeDownloads.contains(objectPath) { return }
            if let entry = self.manifest[objectPath],
               FileManager.default.fileExists(atPath: entry.localPath) { return }
            if self.activeDownloads.count >= self.maxConcurrent { return }

            if !self.hasEnoughStorage() {
                self.evictOldest(count: 1)
                if !self.hasEnoughStorage() {
                    print("[MediaManager] Low storage, skipping: \(objectPath)")
                    return
                }
            }

            self.activeDownloads.insert(objectPath)

            let filename = self.objectPathToFilename(objectPath)
            let filePath = self.mediaDir.appendingPathComponent(filename)
            let tempPath = self.mediaDir.appendingPathComponent(filename + ".tmp")

            guard let url = URL(string: signedUrl) else {
                self.activeDownloads.remove(objectPath)
                return
            }

            let task = URLSession.shared.downloadTask(with: url) { [weak self] tempUrl, response, error in
                guard let self = self else { return }

                defer { self.activeDownloads.remove(objectPath) }

                if let error = error {
                    print("[MediaManager] Download failed for \(objectPath): \(error.localizedDescription)")
                    return
                }

                guard let tempUrl = tempUrl,
                      let httpResponse = response as? HTTPURLResponse,
                      httpResponse.statusCode == 200 else {
                    return
                }

                do {
                    if FileManager.default.fileExists(atPath: filePath.path) {
                        try FileManager.default.removeItem(at: filePath)
                    }
                    try FileManager.default.moveItem(at: tempUrl, to: filePath)

                    let attrs = try FileManager.default.attributesOfItem(atPath: filePath.path)
                    let fileSize = (attrs[.size] as? Int64) ?? 0

                    self.manifest[objectPath] = MediaEntry(
                        localPath: filePath.path,
                        size: fileSize,
                        downloadedAt: Date().timeIntervalSince1970 * 1000,
                        lastAccessed: Date().timeIntervalSince1970 * 1000
                    )
                    self.saveManifest()

                    let protocolUrl = "local-media://" + (filePath.path.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? filePath.path)
                    completion(protocolUrl)
                } catch {
                    print("[MediaManager] Failed to save downloaded file: \(error.localizedDescription)")
                    try? FileManager.default.removeItem(at: tempPath)
                }
            }
            task.resume()
        }
    }

    func deleteMedia(objectPath: String) {
        guard let entry = manifest[objectPath] else { return }
        try? FileManager.default.removeItem(atPath: entry.localPath)
        manifest.removeValue(forKey: objectPath)
        saveManifest()
    }

    func deleteAllMedia() {
        for (_, entry) in manifest {
            try? FileManager.default.removeItem(atPath: entry.localPath)
        }
        manifest = [:]
        saveManifest()
    }

    func getStorageInfo() -> String {
        var usedBytes: Int64 = 0
        var totalFiles = 0
        for entry in manifest.values {
            usedBytes += entry.size
            totalFiles += 1
        }

        var freeBytes: UInt64 = 0
        var totalSpace: UInt64 = 0
        do {
            let attrs = try FileManager.default.attributesOfFileSystem(forPath: mediaDir.path)
            freeBytes = (attrs[.systemFreeSize] as? UInt64) ?? 0
            totalSpace = (attrs[.systemSize] as? UInt64) ?? 0
        } catch {}

        return "{\"usedBytes\":\(usedBytes),\"freeBytes\":\(freeBytes),\"totalSpace\":\(totalSpace),\"totalFiles\":\(totalFiles)}"
    }

    private func hasEnoughStorage() -> Bool {
        do {
            let attrs = try FileManager.default.attributesOfFileSystem(forPath: mediaDir.path)
            let freeBytes = (attrs[.systemFreeSize] as? UInt64) ?? 0
            return freeBytes > lowStorageThreshold
        } catch {
            return true
        }
    }

    private func evictOldest(count: Int) {
        let sorted = manifest.sorted { ($0.value.lastAccessed) < ($1.value.lastAccessed) }
        let toEvict = sorted.prefix(count)
        for (key, entry) in toEvict {
            try? FileManager.default.removeItem(atPath: entry.localPath)
            manifest.removeValue(forKey: key)
        }
        saveManifest()
    }

    func cleanupOrphans() {
        do {
            let files = try FileManager.default.contentsOfDirectory(at: mediaDir, includingPropertiesForKeys: nil)
            let knownPaths = Set(manifest.values.map { $0.localPath })

            for file in files {
                if !knownPaths.contains(file.path) && !file.lastPathComponent.hasSuffix(".tmp") {
                    try? FileManager.default.removeItem(at: file)
                }
            }

            var changed = false
            for (key, entry) in manifest {
                if !FileManager.default.fileExists(atPath: entry.localPath) {
                    manifest.removeValue(forKey: key)
                    changed = true
                }
            }
            if changed { saveManifest() }
        } catch {
            print("[MediaManager] cleanupOrphans failed: \(error.localizedDescription)")
        }
    }

    func cleanup() {
        saveManifest()
    }
}

File: linux/media-manager.js
const path = require('path');
const fs = require('fs');
const https = require('https');
const http = require('http');

const MANIFEST_FILE = 'media-manifest.json';
const MEDIA_DIR = 'media';
const MAX_CONCURRENT = 3;
const LOW_STORAGE_THRESHOLD = 100 * 1024 * 1024;

class MediaManager {
  constructor(userDataPath) {
    this.userDataPath = userDataPath;
    this.mediaDir = path.join(userDataPath, MEDIA_DIR);
    this.manifestPath = path.join(userDataPath, MANIFEST_FILE);
    this.manifest = {};
    this.activeDownloads = new Set();
    this.webContents = null;

    if (!fs.existsSync(this.mediaDir)) {
      fs.mkdirSync(this.mediaDir, { recursive: true });
    }

    this.loadManifest();
  }

  setWebContents(wc) {
    this.webContents = wc;
  }

  loadManifest() {
    try {
      if (fs.existsSync(this.manifestPath)) {
        this.manifest = JSON.parse(fs.readFileSync(this.manifestPath, 'utf-8'));
      }
    } catch (e) {
      console.error('[MediaManager] Failed to load manifest:', e.message);
      this.manifest = {};
    }
  }

  saveManifest() {
    try {
      fs.writeFileSync(this.manifestPath, JSON.stringify(this.manifest, null, 2), 'utf-8');
    } catch (e) {
      console.error('[MediaManager] Failed to save manifest:', e.message);
    }
  }

  objectPathToFilename(objectPath) {
    return objectPath.replace(/^\/objects\//, '').replace(/[\/\\:*?"<>|]/g, '_');
  }

  getLocalMediaPath(objectPath) {
    const entry = this.manifest[objectPath];
    if (!entry) return '';
    const filePath = entry.localPath;
    if (fs.existsSync(filePath)) {
      entry.lastAccessed = Date.now();
      this.saveManifest();
      return 'local-media://' + encodeURIComponent(filePath);
    }
    delete this.manifest[objectPath];
    this.saveManifest();
    return '';
  }

  downloadMedia(objectPath, signedUrl) {
    if (this.activeDownloads.has(objectPath)) return;
    if (this.manifest[objectPath]) {
      const entry = this.manifest[objectPath];
      if (fs.existsSync(entry.localPath)) return;
    }

    if (this.activeDownloads.size >= MAX_CONCURRENT) return;

    if (!this.hasEnoughStorage()) {
      this.evictOldest(1);
      if (!this.hasEnoughStorage()) {
        console.warn('[MediaManager] Low storage, skipping download:', objectPath);
        return;
      }
    }

    this.activeDownloads.add(objectPath);
    const filename = this.objectPathToFilename(objectPath);
    const filePath = path.join(this.mediaDir, filename);
    const tempPath = filePath + '.tmp';

    const client = signedUrl.startsWith('https') ? https : http;

    const request = client.get(signedUrl, (response) => {
      if (response.statusCode === 301 || response.statusCode === 302) {
        this.activeDownloads.delete(objectPath);
        if (response.headers.location) {
          this.downloadMedia(objectPath, response.headers.location);
        }
        return;
      }

      if (response.statusCode !== 200) {
        this.activeDownloads.delete(objectPath);
        this.notifyFailed(objectPath, `HTTP ${response.statusCode}`);
        return;
      }

      const writeStream = fs.createWriteStream(tempPath);
      response.pipe(writeStream);

      writeStream.on('finish', () => {
        writeStream.close(() => {
          try {
            fs.renameSync(tempPath, filePath);
            const stats = fs.statSync(filePath);
            this.manifest[objectPath] = {
              localPath: filePath,
              size: stats.size,
              downloadedAt: Date.now(),
              lastAccessed: Date.now(),
            };
            this.saveManifest();
            this.activeDownloads.delete(objectPath);
            this.notifyDownloaded(objectPath, filePath);
          } catch (e) {
            this.activeDownloads.delete(objectPath);
            this.notifyFailed(objectPath, e.message);
          }
        });
      });

      writeStream.on('error', (e) => {
        this.activeDownloads.delete(objectPath);
        try { fs.unlinkSync(tempPath); } catch (_) {}
        this.notifyFailed(objectPath, e.message);
      });
    });

    request.on('error', (e) => {
      this.activeDownloads.delete(objectPath);
      this.notifyFailed(objectPath, e.message);
    });

    request.setTimeout(120000, () => {
      request.destroy();
      this.activeDownloads.delete(objectPath);
      try { fs.unlinkSync(tempPath); } catch (_) {}
      this.notifyFailed(objectPath, 'Download timeout');
    });
  }

  deleteMedia(objectPath) {
    const entry = this.manifest[objectPath];
    if (!entry) return false;
    try {
      if (fs.existsSync(entry.localPath)) {
        fs.unlinkSync(entry.localPath);
      }
    } catch (e) {
      console.error('[MediaManager] Delete failed:', e.message);
    }
    delete this.manifest[objectPath];
    this.saveManifest();
    return true;
  }

  deleteAllMedia() {
    const count = Object.keys(this.manifest).length;
    for (const [key, entry] of Object.entries(this.manifest)) {
      try {
        if (fs.existsSync(entry.localPath)) {
          fs.unlinkSync(entry.localPath);
        }
      } catch (e) {
        console.error('[MediaManager] Delete failed for', key, e.message);
      }
    }
    this.manifest = {};
    this.saveManifest();
    return count;
  }

  getStorageInfo() {
    let usedBytes = 0;
    let totalFiles = 0;
    for (const entry of Object.values(this.manifest)) {
      usedBytes += entry.size || 0;
      totalFiles++;
    }

    let freeBytes = 0;
    let totalSpace = 0;
    try {
      if (fs.statfsSync) {
        const stats = fs.statfsSync(this.mediaDir);
        freeBytes = stats.bavail * stats.bsize;
        totalSpace = stats.blocks * stats.bsize;
      }
    } catch (_) {}

    return JSON.stringify({ usedBytes, freeBytes, totalSpace, totalFiles });
  }

  hasEnoughStorage() {
    try {
      if (fs.statfsSync) {
        const stats = fs.statfsSync(this.mediaDir);
        const freeBytes = stats.bavail * stats.bsize;
        return freeBytes > LOW_STORAGE_THRESHOLD;
      }
      return true;
    } catch (_) {
      return true;
    }
  }

  evictOldest(count) {
    const entries = Object.entries(this.manifest)
      .sort((a, b) => (a[1].lastAccessed || 0) - (b[1].lastAccessed || 0));
    const toEvict = entries.slice(0, count);
    for (const [key, entry] of toEvict) {
      try {
        if (fs.existsSync(entry.localPath)) {
          fs.unlinkSync(entry.localPath);
        }
      } catch (_) {}
      delete this.manifest[key];
    }
    this.saveManifest();
  }

  cleanupOrphans() {
    try {
      const files = fs.readdirSync(this.mediaDir);
      const knownPaths = new Set(Object.values(this.manifest).map(e => e.localPath));
      for (const file of files) {
        const fullPath = path.join(this.mediaDir, file);
        if (!knownPaths.has(fullPath) && !file.endsWith('.tmp')) {
          try { fs.unlinkSync(fullPath); } catch (_) {}
        }
      }

      for (const [key, entry] of Object.entries(this.manifest)) {
        if (!fs.existsSync(entry.localPath)) {
          delete this.manifest[key];
        }
      }
      this.saveManifest();
    } catch (e) {
      console.error('[MediaManager] cleanupOrphans failed:', e.message);
    }
  }

  notifyDownloaded(objectPath, localPath) {
    if (this.webContents && !this.webContents.isDestroyed()) {
      const protocolUrl = 'local-media://' + encodeURIComponent(localPath);
      this.webContents.executeJavaScript(
        `if(window.__onMediaDownloaded) window.__onMediaDownloaded(${JSON.stringify(objectPath)}, ${JSON.stringify(protocolUrl)})`
      ).catch(() => {});
    }
  }

  notifyFailed(objectPath, error) {
    if (this.webContents && !this.webContents.isDestroyed()) {
      this.webContents.executeJavaScript(
        `if(window.__onMediaDownloadFailed) window.__onMediaDownloadFailed(${JSON.stringify(objectPath)}, ${JSON.stringify(error)})`
      ).catch(() => {});
    }
  }
}

module.exports = MediaManager;
